### DISCUSSION TITLE
Dependency injection for a larger project

### DISCUSSION THREAD

**Discussion by Gidmark** [spectreconsole/spectre.console#1493] [Q&A]
We are working on a larger project where we have a large amount of services and functionality (different databases etc). Are there any recommended patterns to look for when adding a simple CLI tool for this? I don't want to initiate database connections for commands that simply doesn't need them, as an example. The basic idea with the CLI is to add some help commands and test commands.

**Comment by FrankRay78** [comment]
Hello @Gidmark,

There is a basic example of DI usage located in the Examples solution, see `Program.cs` here: https://github.com/spectreconsole/spectre.console/tree/main/examples/Cli/Injection

Also, there are a couple of comments on various threads that might be helpful too, namely:

> @scosec, if you need a prebuild TypeRegistrar, you can have a look at
> 
> * https://www.nuget.org/packages/Spectre.Console.Extensions.Hosting,
> * https://www.nuget.org/packages/Spectre.Console.Registrars.Microsoft-Di or
> * https://www.nuget.org/packages/Spectre.Console.Registrars.SimpleInjector

From https://github.com/spectreconsole/spectre.console/issues/1110

And a real world, fully worked example of a spectre.console user's app with DI, see:

https://github.com/spectreconsole/spectre.console/pull/1412#issuecomment-1868613727  and the user's full repo here: https://github.com/recyclarr/recyclarr/blob/cdf4b2768602f1f4e988d43ee73d14bc097f37c8/src/Recyclarr.Cli/Program.cs

**Reply by Gidmark** [reply]
First of all, thank you for taking the time to answer! 

Looking at the full repo the answer would be that you **do** initiate all the things (CompositionRoot.Setup()) for all commands. I guess that is logical and simplest, I might rethink and make separate tools for different services to avoid having a mountains of dependencies just because i want to test out a http client for example.

**Reply by FrankRay78** [reply]
Things don't always happen fast here in OSS land, however if you start building something and want an opinion or see a genuinely useful extension pattern to spectre.console @Gidmark, feel free to post the code and I'm sure one of us will provide feedback.  https://github.com/spectreconsole/spectre.console/pull/1412#issuecomment-1868613727 was a genuine collaboration between user and core maintainer team, although the underlying issues were open for quite a while.
